### ISSUE TITLE
Mono case insensitive Dictionary and ConcurrentDictionary performance 

### ISSUE THREAD

**Issue by vlad2010** [mono/mono#20628]
I have spotted that Dictionary and ConcurrentDictionary performance is really slow if those collections are created with StringComparer.CurrentCultureIgnoreCase.

I have verified it with .Net 4.6 ( Win 10 x64) , Mono (on macOS Catalina 6.12.0.93 and Linux Ubuntu 20 Mono 6.8.0.105)
), and .Net Core 3.1 (on macOS Catalina).  

Problem only with Mono.
Search and adding elements in case insensitive Dictionary is about 50 times slower in my application. 
(Didn't tests with deleting elements)

Test application attached.

## Steps to Reproduce
Run compile and run attached source file.
[Program.zip](https://github.com/mono/mono/files/5647887/Program.zip)

### Current Behavior

Approximate numbers are the following: 
Run with case insensitive dictionary Mono : about 50-60 sec   <------- 
Run with case sensitive dictionary Mono : less than 1 sec 

Run with case insensitive dictionary .Net and .Net Core  : less than 1 sec 
Run with case sensitive dictionary .Net and .Net Core : less than 1 sec 

### Expected Behavior

The same case insensitive dictionary performance like on .Net and .Net Core 

## On which platforms did you notice this

[x ] macOS
[x] Linux
[ ] Windows

**Version Used**:
on macOS Catalina 6.12.0.93
on Linux Ubuntu 20 Mono 6.8.0.105

**Comment by Therzok** [comment]
Ran a set of benchmarks locally, it seems that StringComparer.Equals and StringComparer.GetHashCode are not the issue.

Modifying your benchmark a bit so it actually has the same data spread across all dictionaries and set the number of searches to 1, since it's enough to demonstrate the problem:

https://gist.github.com/Therzok/2ca82957ab286156247f5e46ffb73b2c

The values _do_ indeed seem off:
```
Time to generate arrays: 00:00.9396795
Time to add to ConcurrentDictionary case sensitive: 00:00.1242398
Time to add to Dictionary case sensitive: 00:00.0865325
Time to add to ConcurrentDictionary CurrentCultureIgnoreCase: 00:12.8525969
Time to add to Dictionary CurrentCultureIgnoreCase: 00:12.9393510
Search started. Key:Olc4sGMMd4
Time to run 10000 iterations. Key:Olc4sGMMd4 Case sensitive ConcurrentDictionary.: 00:00.0035142
Time to run 10000 iterations. Key:Olc4sGMMd4 Case sensitive Dictionary.: 00:00.0032220
Time to run 10000 iterations. Key:Olc4sGMMd4 Case INSENSITIVE ConcurrentDictionary.: 00:01.2892603
Time to run 10000 iterations. Key:Olc4sGMMd4 Case INSENSITIVE Dictionary: 00:01.2570154
```

Looking at the dictionaries themselves, there is:

```
m_ConcDictCaseSensitive._tables._buckets.Length -> 286717
m_ConcDictIgnoreCase._tables._buckets.Length -> 573437
m_DictCaseSensitive._buckets.Length -> 324449
m_DictIgnoreCase._buckets.Length -> 324449
```

I've been running the app multiple times, and it seems like it could be related to the amount of collisions caused by the strings. It seems like mono is not building with `FEATURE_CORECLR` or `FEATURE_RANDOMIZED_STRING_HASHING` here.

A quick profile points to TryInsert taking up most of the time being spent, with GetHashCode taking up around 4%.

<details>
<summary>Benchmarks for string methods</summary>

```
// * Summary *

BenchmarkDotNet=v0.12.0, OS=macOS 10.15.7 (19H2) [Darwin 19.6.0]
Intel Core i9-9880H CPU 2.30GHz, 1 CPU, 16 logical and 8 physical cores
  [Host]     : Mono 6.12.0.113 (2020-02/4fdfb5b1fd5 Mon), X64 
  DefaultJob : Mono 6.12.0.113 (2020-02/4fdfb5b1fd5 Mon), X64 
```


| Method |     ComparisonString |           Comparison |         Mean |      Error |     StdDev |       Median | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------- |--------------------- |--------------------- |-------------:|-----------:|-----------:|-------------:|------:|------:|------:|----------:|
|  Bench |                    a |       CurrentCulture |   281.167 ns |  1.5343 ns |  1.2812 ns |   280.826 ns |     - |     - |     - |         - |
|        |                      |                      |              |            |            |              |       |       |       |           |
|  Bench |                    a | Curre(...)eCase [24] |   293.780 ns |  1.2645 ns |  1.1209 ns |   293.450 ns |     - |     - |     - |         - |
|        |                      |                      |              |            |            |              |       |       |       |           |
|  Bench |                    a |     InvariantCulture |   250.125 ns |  0.7202 ns |  0.5623 ns |   250.201 ns |     - |     - |     - |         - |
|        |                      |                      |              |            |            |              |       |       |       |           |
|  Bench |                    a | Invar(...)eCase [26] |   265.071 ns |  3.4391 ns |  3.2170 ns |   263.853 ns |     - |     - |     - |         - |
|        |                      |                      |              |            |            |              |       |       |       |           |
|  Bench |                    a |              Ordinal |     3.666 ns |  0.0462 ns |  0.0409 ns |     3.656 ns |     - |     - |     - |         - |
|        |                      |                      |              |            |            |              |       |       |       |           |
|  Bench |                    a |    OrdinalIgnoreCase |     3.503 ns |  0.0116 ns |  0.0103 ns |     3.503 ns |     - |     - |     - |         - |
|        |                      |                      |              |            |            |              |       |       |       |           |
|  Bench |  aaaa(...)aaaa [200] |       CurrentCulture |   736.826 ns | 11.1434 ns | 10.4236 ns |   733.041 ns |     - |     - |     - |         - |
|        |                      |                      |              |            |            |              |       |       |       |           |
|  Bench |  aaaa(...)aaaa [200] | Curre(...)eCase [24] |   739.456 ns |  2.0726 ns |  1.7307 ns |   740.000 ns |     - |     - |     - |         - |
|        |                      |                      |              |            |            |              |       |       |       |           |
|  Bench |  aaaa(...)aaaa [200] |     InvariantCulture |   696.561 ns |  2.2539 ns |  1.8821 ns |   695.592 ns |     - |     - |     - |         - |
|        |                      |                      |              |            |            |              |       |       |       |           |
|  Bench |  aaaa(...)aaaa [200] | Invar(...)eCase [26] |   708.468 ns |  1.9997 ns |  1.5612 ns |   708.415 ns |     - |     - |     - |         - |
|        |                      |                      |              |            |            |              |       |       |       |           |
|  Bench |  aaaa(...)aaaa [200] |              Ordinal |     3.651 ns |  0.0115 ns |  0.0102 ns |     3.650 ns |     - |     - |     - |         - |
|        |                      |                      |              |            |            |              |       |       |       |           |
|  Bench |  aaaa(...)aaaa [200] |    OrdinalIgnoreCase |     3.522 ns |  0.0162 ns |  0.0152 ns |     3.523 ns |     - |     - |     - |         - |
|        |                      |                      |              |            |            |              |       |       |       |           |
|  Bench | aaaa(...)aaaa [1000] |       CurrentCulture | 2,322.160 ns |  9.3527 ns |  8.7485 ns | 2,320.344 ns |     - |     - |     - |         - |
|        |                      |                      |              |            |            |              |       |       |       |           |
|  Bench | aaaa(...)aaaa [1000] | Curre(...)eCase [24] | 2,386.721 ns | 47.0818 ns | 67.5233 ns | 2,380.618 ns |     - |     - |     - |         - |
|        |                      |                      |              |            |            |              |       |       |       |           |
|  Bench | aaaa(...)aaaa [1000] |     InvariantCulture | 2,281.712 ns |  6.8904 ns |  6.1081 ns | 2,281.547 ns |     - |     - |     - |         - |
|        |                      |                      |              |            |            |              |       |       |       |           |
|  Bench | aaaa(...)aaaa [1000] | Invar(...)eCase [26] | 2,301.696 ns | 13.8376 ns | 12.9437 ns | 2,295.841 ns |     - |     - |     - |         - |
|        |                      |                      |              |            |            |              |       |       |       |           |
|  Bench | aaaa(...)aaaa [1000] |              Ordinal |     3.656 ns |  0.0132 ns |  0.0110 ns |     3.653 ns |     - |     - |     - |         - |
|        |                      |                      |              |            |            |              |       |       |       |           |
|  Bench | aaaa(...)aaaa [1000] |    OrdinalIgnoreCase |     3.517 ns |  0.0222 ns |  0.0197 ns |     3.520 ns |     - |     - |     - |         - |
|        |                      |                      |              |            |            |              |       |       |       |           |
|  Bench | aaaa(...)aaaa [2000] |       CurrentCulture | 4,277.063 ns |  8.9927 ns |  7.0209 ns | 4,276.947 ns |     - |     - |     - |         - |
|        |                      |                      |              |            |            |              |       |       |       |           |
|  Bench | aaaa(...)aaaa [2000] | Curre(...)eCase [24] | 4,313.539 ns | 19.0914 ns | 16.9240 ns | 4,310.311 ns |     - |     - |     - |         - |
|        |                      |                      |              |            |            |              |       |       |       |           |
|  Bench | aaaa(...)aaaa [2000] |     InvariantCulture | 4,248.358 ns | 16.4931 ns | 15.4277 ns | 4,243.898 ns |     - |     - |     - |         - |
|        |                      |                      |              |            |            |              |       |       |       |           |
|  Bench | aaaa(...)aaaa [2000] | Invar(...)eCase [26] | 4,264.170 ns | 12.7953 ns | 10.6847 ns | 4,264.756 ns |     - |     - |     - |         - |
|        |                      |                      |              |            |            |              |       |       |       |           |
|  Bench | aaaa(...)aaaa [2000] |              Ordinal |   917.194 ns | 18.5787 ns | 26.6450 ns |   902.348 ns |     - |     - |     - |         - |
|        |                      |                      |              |            |            |              |       |       |       |           |
|  Bench | aaaa(...)aaaa [2000] |    OrdinalIgnoreCase | 2,628.256 ns |  9.0801 ns |  8.0493 ns | 2,629.105 ns |     - |     - |     - |         - |
|        |                      |                      |              |            |            |              |       |       |       |           |
|  Bench | aaaa(...)aaaa [4000] |       CurrentCulture | 4,318.666 ns | 36.7905 ns | 32.6138 ns | 4,308.090 ns |     - |     - |     - |         - |
|        |                      |                      |              |            |            |              |       |       |       |           |
|  Bench | aaaa(...)aaaa [4000] | Curre(...)eCase [24] | 4,330.560 ns | 28.9659 ns | 27.0947 ns | 4,319.003 ns |     - |     - |     - |         - |
|        |                      |                      |              |            |            |              |       |       |       |           |
|  Bench | aaaa(...)aaaa [4000] |     InvariantCulture | 4,272.618 ns | 13.8407 ns | 10.8059 ns | 4,275.258 ns |     - |     - |     - |         - |
|        |                      |                      |              |            |            |              |       |       |       |           |
|  Bench | aaaa(...)aaaa [4000] | Invar(...)eCase [26] | 4,273.311 ns | 17.1204 ns | 14.2963 ns | 4,267.216 ns |     - |     - |     - |         - |
|        |                      |                      |              |            |            |              |       |       |       |           |
|  Bench | aaaa(...)aaaa [4000] |              Ordinal |     3.655 ns |  0.0226 ns |  0.0189 ns |     3.651 ns |     - |     - |     - |         - |
|        |                      |                      |              |            |            |              |       |       |       |           |
|  Bench | aaaa(...)aaaa [4000] |    OrdinalIgnoreCase |     3.508 ns |  0.0238 ns |  0.0211 ns |     3.509 ns |     - |     - |     - |         - |

And for GetHashCode:

```
// * Summary *

BenchmarkDotNet=v0.12.0, OS=macOS 10.15.7 (19H2) [Darwin 19.6.0]
Intel Core i9-9880H CPU 2.30GHz, 1 CPU, 16 logical and 8 physical cores
  [Host]     : Mono 6.12.0.113 (2020-02/4fdfb5b1fd5 Mon), X64 
  DefaultJob : Mono 6.12.0.113 (2020-02/4fdfb5b1fd5 Mon), X64 
```

| Method |       HashCodeString |             Comparer |          Mean |        Error |       StdDev |        Median |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------- |--------------------- |--------------------- |--------------:|-------------:|-------------:|--------------:|-------:|------:|------:|----------:|
|  Bench |                    a |       CurrentCulture |     364.02 ns |     9.755 ns |     8.647 ns |     360.70 ns | 0.1445 |     - |     - |         - |
|  Bench |                    a | Curre(...)eCase [24] |     374.39 ns |     1.035 ns |     0.968 ns |     374.21 ns | 0.1445 |     - |     - |         - |
|  Bench |                    a |     InvariantCulture |     352.72 ns |     1.593 ns |     1.490 ns |     352.25 ns | 0.1445 |     - |     - |         - |
|  Bench |                    a | Invar(...)eCase [26] |     358.61 ns |     0.939 ns |     0.784 ns |     358.50 ns | 0.1445 |     - |     - |         - |
|  Bench |                    a |              Ordinal |      11.05 ns |     0.031 ns |     0.029 ns |      11.05 ns |      - |     - |     - |         - |
|  Bench |                    a |    OrdinalIgnoreCase |      68.51 ns |     0.226 ns |     0.200 ns |      68.48 ns |      - |     - |     - |         - |
|  Bench |  aaaa(...)aaaa [500] |       CurrentCulture |  55,178.50 ns |   681.518 ns |   637.492 ns |  54,896.04 ns | 0.9766 |     - |     - |         - |
|  Bench |  aaaa(...)aaaa [500] | Curre(...)eCase [24] |  60,518.84 ns | 1,206.797 ns | 2,465.166 ns |  59,227.45 ns | 0.9766 |     - |     - |         - |
|  Bench |  aaaa(...)aaaa [500] |     InvariantCulture |  52,681.68 ns |   640.762 ns |   568.019 ns |  52,519.73 ns | 0.9766 |     - |     - |         - |
|  Bench |  aaaa(...)aaaa [500] | Invar(...)eCase [26] |  55,818.75 ns |   191.122 ns |   159.596 ns |  55,822.47 ns | 0.9766 |     - |     - |         - |
|  Bench |  aaaa(...)aaaa [500] |              Ordinal |     571.44 ns |     2.793 ns |     2.612 ns |     570.40 ns |      - |     - |     - |         - |
|  Bench |  aaaa(...)aaaa [500] |    OrdinalIgnoreCase |   4,199.82 ns |    14.159 ns |    13.245 ns |   4,196.56 ns |      - |     - |     - |         - |
|  Bench | aaaa(...)aaaa [1000] |       CurrentCulture | 109,548.77 ns |   361.718 ns |   338.352 ns | 109,411.05 ns | 1.5869 |     - |     - |         - |
|  Bench | aaaa(...)aaaa [1000] | Curre(...)eCase [24] | 116,481.10 ns |   566.736 ns |   473.250 ns | 116,326.72 ns | 1.5869 |     - |     - |         - |
|  Bench | aaaa(...)aaaa [1000] |     InvariantCulture | 103,982.41 ns |   740.369 ns |   692.542 ns | 103,925.72 ns | 1.5869 |     - |     - |         - |
|  Bench | aaaa(...)aaaa [1000] | Invar(...)eCase [26] | 110,733.01 ns |   402.095 ns |   356.447 ns | 110,643.18 ns | 1.5869 |     - |     - |         - |
|  Bench | aaaa(...)aaaa [1000] |              Ordinal |   1,124.85 ns |     3.409 ns |     2.847 ns |   1,125.16 ns |      - |     - |     - |         - |
|  Bench | aaaa(...)aaaa [1000] |    OrdinalIgnoreCase |   8,224.80 ns |    29.456 ns |    24.597 ns |   8,221.67 ns |      - |     - |     - |         - |

</details>

**Comment by Therzok** [comment]
Also, performance characteristics for all the string comparison kinds:

```
Time to generate arrays: 00:00.5650141
Looking for gY7tQwOJUB 50000 times

Ordinal
Add: 00:00.0200006
Find: 00:00.0016044

OrdinalIgnoreCase
Add: 00:00.0833054
Find: 00:00.0092845

InvariantCulture
Add: 00:04.1738108
Find: 00:01.7652592

InvariantCultureIgnoreCase
Add: 00:06.1958463
Find: 00:03.1067060

CurrentCulture
Add: 00:04.1680592
Find: 00:01.8126717

CurrentCultureIgnoreCase
Add: 00:06.0390990
Find: 00:03.2448244
```

**Comment by vlad2010** [comment]
Hi,  Therzok   Thank you for your investigation. Where do you think the problem is? Is it possible to try to fix the problem to a person without much knowledge in the mono code base?

**Comment by Therzok** [comment]
Technically, it's up to the mono folks. The problem is missing `FEATURE_RANDOMIZED_STRING_HASHING` in mono, so the input strings are causing a lot of hashcode collisions, so they end up taking a long time to find because they're bucketed together.

**Comment by Therzok** [comment]
cc @marek-safar @lambdageek is there anything we can do short-term? This might impact VSMac too.

**Comment by marek-safar** [comment]
I don't think this is related to FEATURE_RANDOMIZED_STRING_HASHING which is used only for developers not to take a dependency on hastable items ordering.

Culture-specific operations are slow and it's recommended to use Ordinal based methods if possible.

**Comment by Therzok** [comment]
@marek-safar but then why does .netcore not exhibit this issue?

**Comment by Therzok** [comment]
Would switching string's GetHashCode implementation to call into the dotnet one fix it?

https://source.dot.net/#System.Private.CoreLib/String.Comparison.cs,0a17bbac4851d0d4

**Comment by marek-safar** [comment]
No, that code is not used (that's the fast path for ordinar cases). You need to fix 

https://github.com/mono/mono/blob/master/mcs/class/corlib/corefx/CompareInfo.cs#L186-L189
